File: wire-schema/src/main/java/com/squareup/wire/schema/Pruner.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema;

import com.google.common.collect.ImmutableList;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Creates a new schema that contains only the types selected by an identifier set, including their
 * transitive dependencies.
 */
final class Pruner {
  final Schema schema;
  final IdentifierSet identifierSet;
  final MarkSet marks;

  /**
   * {@link ProtoType types} and {@link ProtoMember members} whose immediate dependencies have not
   * yet been visited.
   */
  final Deque<Object> queue;

  Pruner(Schema schema, IdentifierSet identifierSet) {
    this.schema = schema;
    this.identifierSet = identifierSet;
    this.marks = new MarkSet(identifierSet);
    this.queue = new ArrayDeque<>();
  }

  public Schema prune() {
    markRoots();
    markReachable();

    ImmutableList<ProtoFile> retained = retainImports(retainAll(schema, marks));

    return new Schema(retained);
  }

  private ImmutableList<ProtoFile> retainAll(Schema schema, MarkSet marks) {
    ImmutableList.Builder<ProtoFile> retained = ImmutableList.builder();
    for (ProtoFile protoFile : schema.protoFiles()) {
      retained.add(protoFile.retainAll(schema, marks));
    }
    return retained.build();
  }

  private ImmutableList<ProtoFile> retainImports(ImmutableList<ProtoFile> protoFiles) {
    ImmutableList.Builder<ProtoFile> retained = ImmutableList.builder();
    for (ProtoFile protoFile : protoFiles) {
      retained.add(protoFile.retainImports(protoFiles));
    }
    return retained.build();
  }

  private void markRoots() {
    for (ProtoFile protoFile : schema.protoFiles()) {
      markRoots(protoFile);
    }
  }

  private void markRoots(ProtoFile protoFile) {
    for (Type type : protoFile.types()) {
      markRootsIncludingNested(type);
    }
    for (Service service : protoFile.services()) {
      markRoots(service.type());
    }
  }

  private void markRootsIncludingNested(Type type) {
    markRoots(type.type());

    for (Type nested : type.nestedTypes()) {
      markRootsIncludingNested(nested);
    }
  }

  private void markRoots(ProtoType protoType) {
    if (identifierSet.includes(protoType)) {
      marks.root(protoType);
      queue.add(protoType);
      return;
    }

    // The top-level type isn't a root, search for root members inside.
    for (Object reachable : reachableObjects(protoType)) {
      if (reachable instanceof ProtoMember) {
        ProtoMember member = (ProtoMember) reachable;
        if (identifierSet.includes(member)) {
          marks.root(member);
          marks.mark(member.type()); // Consider this type as visited.
          queue.add(member);
        }
      }
    }
  }

  /**
   * Mark everything transitively reachable from the queue, adding to the queue whenever a reachable
   * object brings along more reachable objects.
   */
  private void markReachable() {
    for (Object root; (root = queue.poll()) != null;) {
      List<Object> reachableMembers = reachableObjects(root);

      for (Object reachable : reachableMembers) {
        if (reachable instanceof ProtoType) {
          if (marks.mark((ProtoType) reachable)) {
            queue.add(reachable);
          }
        } else if (reachable instanceof ProtoMember) {
          if (marks.mark((ProtoMember) reachable)) {
            queue.add(reachable);
          }
        } else if (reachable == null) {
          // Skip nulls.
          // TODO(jwilson): create a dedicated UNLINKED type as a placeholder.
        } else {
          throw new IllegalStateException("unexpected object: " + reachable);
        }
      }
    }
  }

  /**
   * Returns everything reachable from {@code root} when traversing the graph. The returned list
   * contains instances of type {@link ProtoMember} and {@link ProtoType}.
   *
   * @param root either a {@link ProtoMember} or {@link ProtoType}.
   */
  private List<Object> reachableObjects(Object root) {
    List<Object> result = new ArrayList<>();
    Options options;

    if (root instanceof ProtoMember) {
      ProtoMember protoMember = (ProtoMember) root;
      String member = ((ProtoMember) root).member();
      Type type = schema.getType(protoMember.type());
      Service service = schema.getService(protoMember.type());

      if (type instanceof MessageType) {
        Field field = ((MessageType) type).field(member);
        if (field == null) {
          field = ((MessageType) type).extensionField(member);
        }
        if (field == null) {
          throw new IllegalStateException("unexpected member: " + member);
        }
        result.add(field.type());
        options = field.options();
      } else if (type instanceof EnumType) {
        EnumConstant constant = ((EnumType) type).constant(member);
        if (constant == null) {
          throw new IllegalStateException("unexpected member: " + member);
        }
        options = constant.getOptions();
      } else if (service != null) {
        Rpc rpc = service.rpc(member);
        if (rpc == null) {
          throw new IllegalStateException("unexpected rpc: " + member);
        }
        result.add(rpc.requestType());
        result.add(rpc.responseType());
        options = rpc.options();
      } else {
        throw new IllegalStateException("unexpected member: " + member);
      }
    } else if (root instanceof ProtoType) {
      ProtoType protoType = (ProtoType) root;

      if (protoType.isMap()) {
        result.add(protoType.keyType());
        result.add(protoType.valueType());
        return result;
      }

      if (protoType.isScalar()) {
        return result; // Skip scalar types.
      }

      Type type = schema.getType(protoType);
      Service service = schema.getService(protoType);

      if (type instanceof MessageType) {
        options = type.options();
        MessageType messageType = (MessageType) type;
        for (Field field : messageType.declaredFields()) {
          result.add(ProtoMember.get(protoType, field.name()));
        }
        for (Field field : messageType.extensionFields()) {
          result.add(ProtoMember.get(protoType, field.qualifiedName()));
        }
        for (OneOf oneOf : messageType.oneOfs()) {
          for (Field field : oneOf.fields()) {
            result.add(ProtoMember.get(protoType, field.name()));
          }
        }
      } else if (type instanceof EnumType) {
        options = type.options();
        EnumType wireEnum = (EnumType) type;
        for (EnumConstant constant : wireEnum.constants()) {
          result.add(ProtoMember.get(wireEnum.type(), constant.getName()));
        }
      } else if (service != null) {
        options = service.options();
        for (Rpc rpc : service.rpcs()) {
          result.add(ProtoMember.get(service.type(), rpc.name()));
        }
      } else {
        throw new IllegalStateException("unexpected type: " + protoType);
      }
    } else {
      throw new IllegalStateException("unexpected root: " + root);
    }

    for (ProtoMember member : options.fields(identifierSet).values()) {
      // If it's an extension, don't consider the entire enclosing type to be reachable.
      if (isExtensionField(member)) {
        result.add(member.type());
      }
      result.add(member);
    }
    return result;
  }

  private boolean isExtensionField(ProtoMember protoMember) {
    Type type = schema.getType(protoMember.type());
    return type instanceof MessageType && ((MessageType) type).field(protoMember.member()) != null;
  }
}

File: wire-schema/src/main/java/com/squareup/wire/schema/Pruner.kt
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema

import com.squareup.wire.schema.ProtoMember.Companion.get
import java.util.ArrayDeque
import java.util.Deque

/**
 * Creates a new schema that contains only the types selected by an identifier set, including their
 * transitive dependencies.
 */
internal class Pruner(
  private val schema: Schema,
  private val pruningRules: PruningRules
) {
  private val marks = MarkSet(pruningRules)
  /**
   * [types][ProtoType] and [members][ProtoMember] whose immediate dependencies have not
   * yet been visited.
   */
  private val queue: Deque<Any> = ArrayDeque()

  fun prune(): Schema {
    markRoots()
    markReachable()

    val retained = retainImports(retainAll(schema, marks))

    return Schema(retained)
  }

  private fun retainAll(
    schema: Schema,
    marks: MarkSet
  ): List<ProtoFile> {
    return schema.protoFiles.map { protoFile ->
      protoFile.retainAll(schema, marks)
    }
  }

  private fun retainImports(protoFiles: List<ProtoFile>): List<ProtoFile> {
    return protoFiles.map { protoFile ->
      protoFile.retainImports(protoFiles)
    }
  }

  private fun markRoots() {
    for (protoFile in schema.protoFiles) {
      markRoots(protoFile)
    }
  }

  private fun markRoots(protoFile: ProtoFile) {
    for (type in protoFile.types) {
      markRootsIncludingNested(type)
    }
    for (service in protoFile.services) {
      markRoots(service.type())
    }
  }

  private fun markRootsIncludingNested(type: Type) {
    markRoots(type.type!!)

    for (nested in type.nestedTypes) {
      markRootsIncludingNested(nested)
    }
  }

  private fun markRoots(protoType: ProtoType) {
    if (pruningRules.includes(protoType)) {
      marks.root(protoType)
      queue.add(protoType)
      return
    }

    // The top-level type isn't a root, search for root members inside.
    for (reachable in reachableObjects(protoType)) {
      if (reachable !is ProtoMember) continue
      if (!isRetainedVersion(reachable)) continue
      if (pruningRules.includes(reachable)) {
        marks.root(reachable)
        marks.mark(reachable.type) // Consider this type as visited.
        queue.add(reachable)
      }
    }
  }

  /** Returns true if this member survives `since` and `until` pruning. */
  private fun isRetainedVersion(protoMember: ProtoMember): Boolean {
    val member = protoMember.member
    val type = schema.getType(protoMember.type)

    if (type is MessageType) {
      val field = type.field(member) ?: type.extensionField(member)!!
      return pruningRules.isRetainedVersion(field.options)
    }

    return true
  }

  /**
   * Mark everything transitively reachable from the queue, adding to the queue whenever a reachable
   * object brings along more reachable objects.
   */
  private fun markReachable() {
    while (true) {
      val root: Any = queue.poll() ?: break
      val reachableMembers: List<Any?> = reachableObjects(root)

      for (reachable in reachableMembers) {
        when (reachable) {
          is ProtoType -> {
            if (marks.mark(reachable)) {
              queue.add(reachable)
            }
          }

          is ProtoMember -> {
            if (isRetainedVersion(reachable) && marks.mark(reachable)) {
              queue.add(reachable)
            }
          }

          null -> {
            // Skip nulls.
            // TODO(jwilson): create a dedicated UNLINKED type as a placeholder.
          }

          else -> {
            throw IllegalStateException("unexpected object: $reachable")
          }
        }
      }
    }
  }

  /**
   * Returns everything reachable from `root` when traversing the graph. The returned list
   * contains instances of type [ProtoMember] and [ProtoType].
   *
   * @param root either a [ProtoMember] or [ProtoType].
   */
  private fun reachableObjects(root: Any): List<Any?> {
    val result = mutableListOf<Any?>()
    val options: Options
    var fileOptions: Options? = null

    when (root) {
      is ProtoMember -> {
        val member = root.member
        val type = schema.getType(root.type)
        val service = schema.getService(root.type)

        if (type is MessageType) {
          val field = type.field(member) ?: type.extensionField(member)
          checkNotNull(field) { "unexpected member: $member" }
          result.add(field.type)
          options = field.options

        } else if (type is EnumType) {
          val constant = type.constant(member)
              ?: throw IllegalStateException("unexpected member: $member")
          options = constant.options

        } else if (service != null) {
          val rpc = service.rpc(member) ?: throw IllegalStateException("unexpected rpc: $member")
          result.add(rpc.requestType)
          result.add(rpc.responseType)
          options = rpc.options

        } else {
          throw IllegalStateException("unexpected member: $member")
        }
      }

      is ProtoType -> {
        if (root.isMap) {
          result.add(root.keyType)
          result.add(root.valueType)
          return result
        }

        if (root.isScalar) {
          return result // Skip scalar types.
        }

        val type = schema.getType(root)
        val service = schema.getService(root)
        fileOptions = schema.protoFile(root).options

        if (type is MessageType) {
          options = type.options
          for (field in type.declaredFields) {
            result.add(get(root, field.name))
          }
          for (field in type.extensionFields) {
            result.add(get(root, field.qualifiedName))
          }
          for (oneOf in type.oneOfs) {
            for (field in oneOf.fields) {
              result.add(get(root, field.name))
            }
          }

        } else if (type is EnumType) {
          options = type.options
          for (constant in type.constants) {
            result.add(get(type.type, constant.name))
          }

        } else if (service != null) {
          options = service.options()
          for (rpc in service.rpcs()) {
            result.add(get(service.type(), rpc.name))
          }

        } else {
          throw IllegalStateException("unexpected type: $root")
        }
      }
      else -> {
        throw IllegalStateException("unexpected root: $root")
      }
    }

    addOptions(options.fields(pruningRules).values(), result)
    if (fileOptions != null) {
      addOptions(fileOptions.fields(pruningRules).values(), result)
    }

    return result
  }

  private fun addOptions(options: Collection<ProtoMember>, result: MutableList<Any?>) {
    for (member in options) {
      // If it's an extension, don't consider the entire enclosing type to be reachable.
      if (!isExtensionField(member)) {
        result.add(member.type)
      }
      result.add(member)
    }
  }

  private fun isExtensionField(protoMember: ProtoMember): Boolean {
    val type = schema.getType(protoMember.type)
    return type is MessageType && type.extensionField(protoMember.member) != null
  }
}
